test/model/internal_link_test.dart

```diff
@@ -0,0 +1,394 @@
+    for (final testCase in testCases) {
+      final bool expected = testCase.$1;
+      final String description = testCase.$2;
+      final Uri realmUrl = testCase.$3;
+      final Uri url = testCase.$4;
+      test('${expected ? 'accepts': 'rejects'} $description: $url', () {
+        final store = setupStore(realmUrl: realmUrl, streams: streams);
+        final result = parseInternalLink(url, store);
+        check(result != null).equals(expected);
+      });
+    }
+  });
+
+  group('ParseInternalLink', () {
+    final streams = [
+      eg.stream(streamId: 1,   name: 'check'),
+      eg.stream(streamId: 3,   name: 'mobile'),
+      eg.stream(streamId: 5,   name: 'stream'),
+      eg.stream(streamId: 123, name: 'topic'),
+    ];
+
+    group('"/#narrow/stream/<...>" returns expected StreamNarrow', () {
+      const testCases = [
+        ('/#narrow/stream/check',   StreamNarrow(1)),
+        ('/#narrow/stream/stream/', StreamNarrow(5)),
+        ('/#narrow/stream/topic/',  StreamNarrow(123)),
+      ];
+      checkExpectedNarrows(testCases, streams: streams);
+    });
+
+    group('"/#narrow/stream/<...>/topic/<...>" returns expected TopicNarrow', () {
+      const testCases = [
+        ('/#narrow/stream/check/topic/test',                 TopicNarrow(1, 'test')),
+        ('/#narrow/stream/mobile/subject/topic/near/378333', TopicNarrow(3, 'topic')),
+        ('/#narrow/stream/mobile/topic/topic/',              TopicNarrow(3, 'topic')),
+        ('/#narrow/stream/stream/topic/topic/near/1',        TopicNarrow(5, 'topic')),
+        ('/#narrow/stream/stream/subject/topic/near/1',      TopicNarrow(5, 'topic')),
+        ('/#narrow/stream/stream/subject/topic',             TopicNarrow(5, 'topic')),
+      ];
+      checkExpectedNarrows(testCases, streams: streams);
+    });
+
+    group('"/#narrow/dm/<...>" returns expected DmNarrow', () {
+      final expectedNarrow = DmNarrow.withUsers([1, 2],
+        selfUserId: eg.selfUser.userId);
+      final testCases = [
+        ('/#narrow/dm/1,2-group',                        expectedNarrow),
+        ('/#narrow/dm/1,2-group/near/1',                 expectedNarrow),
+        ('/#narrow/dm/a.40b.2Ecom.2Ec.2Ed.2Ecom/near/3', null),
+      ];
+      checkExpectedNarrows(testCases, streams: streams);
+    });
+
+    group('"/#narrow/pm-with/<...>" returns expected DmNarrow', () {
+      final expectedNarrow = DmNarrow.withUsers([1, 2],
+        selfUserId: eg.selfUser.userId);
+      final testCases = [
+        ('/#narrow/pm-with/1,2-group',                        expectedNarrow),
+        ('/#narrow/pm-with/1,2-group/near/1',                 expectedNarrow),
+        ('/#narrow/pm-with/a.40b.2Ecom.2Ec.2Ed.2Ecom/near/3', null),
+      ];
+      checkExpectedNarrows(testCases, streams: streams);
+    });
+
+    group('unexpected link shapes are rejected', () {
+      final testCases = [
+        ('/#narrow/stream/name/topic/',           null),
+        ('/#narrow/stream/name/unknown/operand/', null),
```

Comment:
```suggestion
      final testCases = [
        ('/#narrow/stream/name/topic/',           null), // missing operand
        ('/#narrow/stream/name/unknown/operand/', null), // unknown operator
```

(like the comments in the zulip-mobile version)